client/asset/dcr/dcr.go

```diff
@@ -0,0 +1,1181 @@
+			}
+			size := msgTx.SerializeSize()
+			reqFee := dcr.nfo.FeeRate * uint64(size)
+			if reqFee > remaining {
+				// I can't imagine a scenario where this condition would be true, but
+				// I'd hate to be wrong.
+				dcr.log.Errorf("reached the impossible place. in = %d, out = %d, reqFee = %d, lastFee = %d",
+					totalIn, totalOut, reqFee, fee)
+				return nil, fmt.Errorf("change error")
+			}
+			if fee == reqFee || (reqFee < fee && tried[reqFee] == 1) {
+				// If we've already tried it, we're on our way back up, so this fee is
+				// likely as good as it gets.
+				break
+			}
+			// We must have some room for improvement
+			tried[fee] = 1
+			fee = reqFee
+			changeOutput.Value = int64(remaining - fee)
+			if dexdcr.IsDust(changeOutput, dcr.nfo.FeeRate) {
+				// Another condition that should be impossible, but check anyway in case
+				// the maximum fee was underestimated causing the first check to be
+				// missed.
+				dcr.log.Errorf("reached the impossible place. in = %d, out = %d, reqFee = %d, lastFee = %d",
+					totalIn, totalOut, reqFee, fee)
+				return nil, fmt.Errorf("dust error")
+			}
+			continue
+		}
+	}
+
+	checkHash := msgTx.TxHash()
+	dcr.log.Debugf("%d signature cycles to converge on fees for tx %s", sigCycles, checkHash)
+	txHash, err := dcr.node.SendRawTransaction(msgTx, false)
+	if err != nil {
+		return nil, err
+	}
+	if *txHash != checkHash {
+		return nil, fmt.Errorf("transaction sent, but received unexpected transaction ID back from RPC server. "+
+			"expected %s, got %s", *txHash, checkHash)
+	}
+	if !isDust {
+		dcr.addChange(txHash, uint32(changeIdx))
+	}
+	return msgTx, nil
+}
+
+// For certain dcrutil.Address types.
+type signatureTyper interface {
+	DSA() dcrec.SignatureType
+}
+
+// createSig creates and returns the serialized raw signature and compressed
+// pubkey for a transaction input signature.
+func (dcr *ExchangeWallet) createSig(tx *wire.MsgTx, idx int, pkScript []byte, addr dcrutil.Address) (sig, pubkey []byte, err error) {
+	sigTyper, ok := addr.(signatureTyper)
+	if !ok {
+		return nil, nil, fmt.Errorf("invalid address type")
+	}
+
+	priv, pub, err := dcr.getKeys(addr)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	sigType := sigTyper.DSA()
+	switch sigType {
+	case dcrec.STEcdsaSecp256k1:
+		sig, err = txscript.RawTxInSignature(tx, idx, pkScript, txscript.SigHashAll, priv)
+	default:
+		sig, err = txscript.RawTxInSignatureAlt(tx, idx, pkScript, txscript.SigHashAll, priv, sigType)
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return sig, pub.SerializeCompressed(), nil
+}
+
+// getKeys fetches the private/public key pair for the specified address.
+func (dcr *ExchangeWallet) getKeys(addr dcrutil.Address) (*secp256k1.PrivateKey, *secp256k1.PublicKey, error) {
+	wif, err := dcr.node.DumpPrivKey(addr, chainParams.PrivateKeyID)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	priv, pub := secp256k1.PrivKeyFromBytes(wif.PrivKey.Serialize())
```

Comment:
I believe the `dcrutil.WIF` type is changing in `dcrutil/v3` and the private key is no longer exported, so this is going to need fixing when we upgrade our dcrd dependencies.